File: test/java/brs/web/api/http/handler/IssueAssetTest.java
package brs.web.api.http.handler;

import brs.Attachment;
import brs.Blockchain;
import brs.Signum;
import brs.SignumException;
import brs.Constants;
import brs.common.QuickMocker;
import brs.common.QuickMocker.MockParam;
import brs.fluxcapacitor.FluxCapacitor;
import brs.fluxcapacitor.FluxValues;
import brs.services.ParameterService;
import brs.web.api.http.common.APITransactionManager;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.powermock.core.classloader.annotations.PrepareForTest;
import org.powermock.modules.junit4.PowerMockRunner;

import javax.servlet.http.HttpServletRequest;

import static brs.Constants.*;
import static brs.TransactionType.ColoredCoins.ASSET_ISSUANCE;
import static brs.web.api.http.common.JSONResponses.*;
import static brs.web.api.http.common.Parameters.*;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;
import static org.powermock.api.mockito.PowerMockito.mockStatic;

@RunWith(PowerMockRunner.class)
@PrepareForTest(Signum.class)
public class IssueAssetTest extends AbstractTransactionTest {

  private IssueAsset t;

  private ParameterService mockParameterService;
  private Blockchain mockBlockchain;
  private APITransactionManager apiTransactionManagerMock;

  @Before
  public void setUp() {
    mockParameterService = mock(ParameterService.class);
    mockBlockchain = mock(Blockchain.class);
    apiTransactionManagerMock = mock(APITransactionManager.class);

    t = new IssueAsset(mockParameterService, mockBlockchain, apiTransactionManagerMock);
  }

  @Test
  public void processRequest() throws SignumException {
    final String nameParameter = stringWithLength(MIN_ASSET_NAME_LENGTH + 1);
    final String descriptionParameter = stringWithLength(MAX_ASSET_DESCRIPTION_LENGTH - 1);
    final int decimalsParameter = 4;
    final int quantityQNTParameter = 5;

    final HttpServletRequest req = QuickMocker.httpServletRequest(
        new MockParam(NAME_PARAMETER, nameParameter),
        new MockParam(DESCRIPTION_PARAMETER, descriptionParameter),
        new MockParam(DECIMALS_PARAMETER, decimalsParameter),
        new MockParam(QUANTITY_QNT_PARAMETER, quantityQNTParameter)
    );

    mockStatic(Signum.class);
    final FluxCapacitor fluxCapacitor = QuickMocker.fluxCapacitorEnabledFunctionalities(FluxValues.DIGITAL_GOODS_STORE);
    when(Signum.getFluxCapacitor()).thenReturn(fluxCapacitor);
    doReturn(Constants.FEE_QUANT_SIP3).when(fluxCapacitor).getValue(eq(FluxValues.FEE_QUANT));

    final Attachment.ColoredCoinsAssetIssuance attachment = (Attachment.ColoredCoinsAssetIssuance) attachmentCreatedTransaction(() -> t.processRequest(req), apiTransactionManagerMock);
    assertNotNull(attachment);

    assertEquals(ASSET_ISSUANCE, attachment.getTransactionType());
    assertEquals(nameParameter, attachment.getName());
    assertEquals(descriptionParameter, attachment.getDescription());
    assertEquals(decimalsParameter, attachment.getDecimals());
    assertEquals(descriptionParameter, attachment.getDescription());
  }

  @Test
  public void processRequest_missingName() throws SignumException {
    final HttpServletRequest req = QuickMocker.httpServletRequest();

    assertEquals(MISSING_NAME, t.processRequest(req));
  }

  @Test
  public void processRequest_incorrectAssetNameLength_smallerThanMin() throws SignumException {
    final HttpServletRequest req = QuickMocker.httpServletRequest(
        new MockParam(NAME_PARAMETER, stringWithLength(MIN_ASSET_NAME_LENGTH - 1))
    );

    assertEquals(INCORRECT_ASSET_NAME_LENGTH, t.processRequest(req));
  }

  @Test
  public void processRequest_incorrectAssetNameLength_largerThanMax() throws SignumException {
    final HttpServletRequest req = QuickMocker.httpServletRequest(
        new MockParam(NAME_PARAMETER, stringWithLength(MAX_ASSET_NAME_LENGTH + 1))
    );

    assertEquals(INCORRECT_ASSET_NAME_LENGTH, t.processRequest(req));
  }

  @Test
  public void processRequest_incorrectAssetName() throws SignumException {
    final HttpServletRequest req = QuickMocker.httpServletRequest(
        new MockParam(NAME_PARAMETER, stringWithLength(MIN_ASSET_NAME_LENGTH + 1) + "[")
    );

    assertEquals(INCORRECT_ASSET_NAME, t.processRequest(req));
  }

  @Test
  public void processRequest_incorrectAssetDescription() throws SignumException {
    final HttpServletRequest req = QuickMocker.httpServletRequest(
        new MockParam(NAME_PARAMETER, stringWithLength(MIN_ASSET_NAME_LENGTH + 1)),
        new MockParam(DESCRIPTION_PARAMETER, stringWithLength(MAX_ASSET_DESCRIPTION_LENGTH + 1))
    );

    assertEquals(INCORRECT_ASSET_DESCRIPTION, t.processRequest(req));
  }

  @Test
  public void processRequest_incorrectDecimals_unParsable() throws SignumException {
    final HttpServletRequest req = QuickMocker.httpServletRequest(
        new MockParam(NAME_PARAMETER, stringWithLength(MIN_ASSET_NAME_LENGTH + 1)),
        new MockParam(DESCRIPTION_PARAMETER, stringWithLength(MAX_ASSET_DESCRIPTION_LENGTH - 1)),
        new MockParam(DECIMALS_PARAMETER, "unParsable")
    );

    assertEquals(INCORRECT_DECIMALS, t.processRequest(req));
  }

  @Test
  public void processRequest_incorrectDecimals_negativeNumber() throws SignumException {
    final HttpServletRequest req = QuickMocker.httpServletRequest(
        new MockParam(NAME_PARAMETER, stringWithLength(MIN_ASSET_NAME_LENGTH + 1)),
        new MockParam(DESCRIPTION_PARAMETER, stringWithLength(MAX_ASSET_DESCRIPTION_LENGTH - 1)),
        new MockParam(DECIMALS_PARAMETER, -5)
    );

    assertEquals(INCORRECT_DECIMALS, t.processRequest(req));
  }

  @Test
  public void processRequest_incorrectDecimals_moreThan8() throws SignumException {
    final HttpServletRequest req = QuickMocker.httpServletRequest(
        new MockParam(NAME_PARAMETER, stringWithLength(MIN_ASSET_NAME_LENGTH + 1)),
        new MockParam(DESCRIPTION_PARAMETER, stringWithLength(MAX_ASSET_DESCRIPTION_LENGTH - 1)),
        new MockParam(DECIMALS_PARAMETER, 9)
    );

    assertEquals(INCORRECT_DECIMALS, t.processRequest(req));
  }

}

File: test/java/brs/http/IssueAssetTest.kt
package brs.http

import brs.Attachment
import brs.Blockchain
import brs.Burst
import brs.BurstException
import brs.common.QuickMocker
import brs.common.QuickMocker.MockParam
import brs.fluxcapacitor.FluxCapacitor
import brs.fluxcapacitor.FluxValues
import brs.services.ParameterService
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith

import javax.servlet.http.HttpServletRequest

import brs.Constants.*
import brs.TransactionType.ColoredCoins.ASSET_ISSUANCE
import brs.http.JSONResponses.INCORRECT_ASSET_DESCRIPTION
import brs.http.JSONResponses.INCORRECT_ASSET_NAME
import brs.http.JSONResponses.INCORRECT_ASSET_NAME_LENGTH
import brs.http.JSONResponses.INCORRECT_DECIMALS
import brs.http.JSONResponses.MISSING_NAME
import brs.http.common.Parameters.*
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.whenever
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.runners.JUnit4

@RunWith(JUnit4::class)
class IssueAssetTest : AbstractTransactionTest() {

    private var t: IssueAsset? = null

    private var mockParameterService: ParameterService? = null
    private var mockBlockchain: Blockchain? = null
    private var apiTransactionManagerMock: APITransactionManager? = null

    @Before
    fun setUp() {
        mockParameterService = mock<ParameterService>()
        mockBlockchain = mock<Blockchain>()
        apiTransactionManagerMock = mock<APITransactionManager>()

        t = IssueAsset(mockParameterService!!, mockBlockchain!!, apiTransactionManagerMock!!)
    }

    @Test
    @Throws(BurstException::class)
    fun processRequest() {
        val nameParameter = stringWithLength(MIN_ASSET_NAME_LENGTH + 1)
        val descriptionParameter = stringWithLength(MAX_ASSET_DESCRIPTION_LENGTH - 1)
        val decimalsParameter = 4
        val quantityQNTParameter = 5

        val req = QuickMocker.httpServletRequest(
                MockParam(NAME_PARAMETER, nameParameter),
                MockParam(DESCRIPTION_PARAMETER, descriptionParameter),
                MockParam(DECIMALS_PARAMETER, decimalsParameter),
                MockParam(QUANTITY_QNT_PARAMETER, quantityQNTParameter)
        )

        QuickMocker.fluxCapacitorEnabledFunctionalities(FluxValues.DIGITAL_GOODS_STORE)

        val attachment = attachmentCreatedTransaction({ t!!.processRequest(req) }, apiTransactionManagerMock!!) as Attachment.ColoredCoinsAssetIssuance
        assertNotNull(attachment)

        assertEquals(ASSET_ISSUANCE, attachment.transactionType)
        assertEquals(nameParameter, attachment.name)
        assertEquals(descriptionParameter, attachment.description)
        assertEquals(decimalsParameter.toLong(), attachment.decimals.toLong())
        assertEquals(descriptionParameter, attachment.description)
    }

    @Test
    @Throws(BurstException::class)
    fun processRequest_missingName() {
        val req = QuickMocker.httpServletRequest()

        assertEquals(MISSING_NAME, t!!.processRequest(req))
    }

    @Test
    @Throws(BurstException::class)
    fun processRequest_incorrectAssetNameLength_smallerThanMin() {
        val req = QuickMocker.httpServletRequest(
                MockParam(NAME_PARAMETER, stringWithLength(MIN_ASSET_NAME_LENGTH - 1))
        )

        assertEquals(INCORRECT_ASSET_NAME_LENGTH, t!!.processRequest(req))
    }

    @Test
    @Throws(BurstException::class)
    fun processRequest_incorrectAssetNameLength_largerThanMax() {
        val req = QuickMocker.httpServletRequest(
                MockParam(NAME_PARAMETER, stringWithLength(MAX_ASSET_NAME_LENGTH + 1))
        )

        assertEquals(INCORRECT_ASSET_NAME_LENGTH, t!!.processRequest(req))
    }

    @Test
    @Throws(BurstException::class)
    fun processRequest_incorrectAssetName() {
        val req = QuickMocker.httpServletRequest(
                MockParam(NAME_PARAMETER, stringWithLength(MIN_ASSET_NAME_LENGTH + 1) + "[")
        )

        assertEquals(INCORRECT_ASSET_NAME, t!!.processRequest(req))
    }

    @Test
    @Throws(BurstException::class)
    fun processRequest_incorrectAssetDescription() {
        val req = QuickMocker.httpServletRequest(
                MockParam(NAME_PARAMETER, stringWithLength(MIN_ASSET_NAME_LENGTH + 1)),
                MockParam(DESCRIPTION_PARAMETER, stringWithLength(MAX_ASSET_DESCRIPTION_LENGTH + 1))
        )

        assertEquals(INCORRECT_ASSET_DESCRIPTION, t!!.processRequest(req))
    }

    @Test
    @Throws(BurstException::class)
    fun processRequest_incorrectDecimals_unParsable() {
        val req = QuickMocker.httpServletRequest(
                MockParam(NAME_PARAMETER, stringWithLength(MIN_ASSET_NAME_LENGTH + 1)),
                MockParam(DESCRIPTION_PARAMETER, stringWithLength(MAX_ASSET_DESCRIPTION_LENGTH - 1)),
                MockParam(DECIMALS_PARAMETER, "unParsable")
        )

        assertEquals(INCORRECT_DECIMALS, t!!.processRequest(req))
    }

    @Test
    @Throws(BurstException::class)
    fun processRequest_incorrectDecimals_negativeNumber() {
        val req = QuickMocker.httpServletRequest(
                MockParam(NAME_PARAMETER, stringWithLength(MIN_ASSET_NAME_LENGTH + 1)),
                MockParam(DESCRIPTION_PARAMETER, stringWithLength(MAX_ASSET_DESCRIPTION_LENGTH - 1)),
                MockParam(DECIMALS_PARAMETER, -5L)
        )

        assertEquals(INCORRECT_DECIMALS, t!!.processRequest(req))
    }

    @Test
    @Throws(BurstException::class)
    fun processRequest_incorrectDecimals_moreThan8() {
        val req = QuickMocker.httpServletRequest(
                MockParam(NAME_PARAMETER, stringWithLength(MIN_ASSET_NAME_LENGTH + 1)),
                MockParam(DESCRIPTION_PARAMETER, stringWithLength(MAX_ASSET_DESCRIPTION_LENGTH - 1)),
                MockParam(DECIMALS_PARAMETER, 9L)
        )

        assertEquals(INCORRECT_DECIMALS, t!!.processRequest(req))
    }

}
